internal/service/advancedcluster/model_advanced_cluster_test.go

```diff
@@ -21,6 +23,128 @@ var (
 	advancedClusters = []admin.AdvancedClusterDescription{{StateName: conversion.StringPtr("NOT IDLE")}}
 )
 
+func TestFlattenReplicationSpecs(t *testing.T) {
+	var (
+		regionName         = "EU_WEST_1"
+		providerName       = "AWS"
+		expectedID         = "id1"
+		unexpectedID       = "id2"
+		expectedZoneName   = "z1"
+		unexpectedZoneName = "z2"
+		regionConfigAdmin  = []admin.CloudRegionConfig{{
+			ProviderName: &providerName,
+			RegionName:   &regionName,
+		}}
+		regionConfigTf = map[string]any{
+			"provider_name": "AWS",
+			"region_name":   regionName,
+		}
+		regionConfigTfDiffZone = map[string]any{
+			"provider_name": "AWS",
+			"region_name":   regionName,
+			"zone_name":     unexpectedZoneName,
+		}
+		admin1     = admin.ReplicationSpec{Id: &expectedID, ZoneName: &expectedZoneName, RegionConfigs: &regionConfigAdmin}
+		admin2     = admin.ReplicationSpec{Id: &unexpectedID, ZoneName: &unexpectedZoneName, RegionConfigs: &regionConfigAdmin}
+		testSchema = map[string]*schema.Schema{
+			"project_id": {Type: schema.TypeString},
+		}
+		tf1SameIDSameZone = map[string]any{
+			"id":             expectedID,
+			"num_shards":     1,
+			"region_configs": []any{regionConfigTf},
+			"zone_name":      expectedZoneName,
+		}
+		tf2NoIDSameZone = map[string]any{
+			"id":             nil,
+			"num_shards":     1,
+			"region_configs": []any{regionConfigTf},
+			"zone_name":      expectedZoneName,
+		}
+		tf3NoIDDiffZone = map[string]any{
+			"id":             nil,
+			"num_shards":     1,
+			"region_configs": []any{regionConfigTfDiffZone},
+			"zone_name":      unexpectedZoneName,
+		}
+		tf4diffIDDiffZone = map[string]any{
+			"id":             "unique",
+			"num_shards":     1,
+			"region_configs": []any{regionConfigTfDiffZone},
+			"zone_name":      unexpectedZoneName,
+		}
+	)
+	testCases := map[string]struct {
+		adminSpecs   []admin.ReplicationSpec
+		tfInputSpecs []any
+		expectedLen  int
+	}{
+		"empty admin spec should return empty list": {
+			[]admin.ReplicationSpec{},
+			[]any{tf1SameIDSameZone},
+			0,
+		},
+		"existing id, should match admin": {
+			[]admin.ReplicationSpec{admin1},
+			[]any{tf1SameIDSameZone},
+			1,
+		},
+		"existing different id, should change to admin spec": {
+			[]admin.ReplicationSpec{admin1},
+			[]any{tf4diffIDDiffZone},
+			1,
+		},
+		"missing id, should be set when zone_name matches": {
+			[]admin.ReplicationSpec{admin1},
+			[]any{tf2NoIDSameZone},
+			1,
+		},
+		"missing id and diff zone, should change to admin spec": {
+			[]admin.ReplicationSpec{admin1},
+			[]any{tf3NoIDDiffZone},
+			1,
+		},
```

Comment:
Just for my understanding, in this case if there is no match the values from the api prevail right?

Reply:
Correct